proof-server/tests/integration_tests.rs

```
@@ -339,6 +559,216 @@ async fn test_prove_tx_should_fail_on_json() {
+async fn test_prove_should_fail_with_double_zk_config() {
+    setup_test(function_name!());
+
+    let response = HTTP_CLIENT
+        .post(format!("{}/prove", get_host_and_port()))
+        .body(serialized_body_with_double_zk_config().await)
+        .send()
+        .await
+        .unwrap();
+
+    log::info!("Response code: {:?}", response.status());
+    assert_eq!(response.status(), 400);
+    let resp_text = response.text().await.unwrap();
+    assert!(resp_text.contains("expected header tag"));
+}
+
+// Negative test: `/prove` must reject payloads where ZK Config values are value-swapped.
+// Given:
+// - A valid request `(tx, {A -> zkA, B -> zkB, ...})`.
+// - We swap the proving material so that `{A -> zkB, B -> zkA}`.
+// When: The malformed payload is POSTed to `/prove`.
+// Then: The server responds with `400 Bad Request`.
+#[named]
+async fn test_prove_should_fail_with_swapped_zk_config_values() {
+    setup_test(function_name!());
+
+    if let Some(payload) = serialized_body_with_swapped_zk_config_values().await {
+        let response = HTTP_CLIENT
+            .post(format!("{}/prove", get_host_and_port()))
+            .body(payload)
+            .send()
+            .await
+            .unwrap();
+
+        log::info!("Response code: {:?}", response.status());
+        assert_eq!(response.status(), 400);
+        let resp_text = response.text().await.unwrap();
+        assert!(resp_text.contains("failed to find proving key"));
+    }
+}
+
+// Negative test: `/prove` must reject payloads whose ZK Config contains an
+// entry keyed by a circuit ID that is not part of the transaction.
+// Given: A valid request `(tx, {A -> zkA, ...})`.
+// When: We replace one key with `C_wrong` while keeping the proving material.
+// Then: The server responds with `400 Bad Request`.
+#[named]
+async fn test_prove_should_fail_with_wrong_circuit_id() {
+    setup_test(function_name!());
+
+    if let Some(payload) = serialized_body_with_wrong_circuit_id().await {
```

This will silently pass the test if `serialized_body_with_swapped_zk_config_values` fails 😢